auth/clients.go

```diff
@@ -2,56 +2,122 @@ package auth
 
 import (
 	"fmt"
-	"os"
+	"strings"
 
+	"github.com/codegangsta/cli"
+	"github.com/jrperritt/rack/util"
 	"github.com/rackspace/gophercloud"
 	"github.com/rackspace/gophercloud/rackspace"
 )
 
-// NewClient creates and returns a Rackspace client for the given service.
-func NewClient(t string) *gophercloud.ServiceClient {
-	var err error
-	ao, region, err := authMethod()
-	if err != nil {
-		fmt.Printf("Error retrieving authentication credentials: %s\n", err)
+// reauthFunc is what the ServiceClient uses to re-authenticate.
+func reauthFunc(pc *gophercloud.ProviderClient, ao gophercloud.AuthOptions) func() error {
+	return func() error {
+		return rackspace.AuthenticateV2(pc, ao)
 	}
+}
+
+// NewClient creates and returns a Rackspace client for the given service.
+func NewClient(c *cli.Context, serviceType string) (*gophercloud.ServiceClient, error) {
+	// get the user's authentication credentials
+	ao, region := Credentials(c)
+	// upper-case the region
+	region = strings.ToUpper(region)
+	// allow Gophercloud to re-authenticate
+	ao.AllowReauth = true
+	// if the user didn't provide an auth URL, default to the Rackspace US endpoint
 	if ao.IdentityEndpoint == "" {
 		ao.IdentityEndpoint = rackspace.RackspaceUSIdentity
 	}
-	pc, err := rackspace.AuthenticatedClient(ao)
-	if err != nil {
-		fmt.Printf("Error creating ProviderClient: %s\n", err)
-		os.Exit(1)
-	}
-	var sc *gophercloud.ServiceClient
-	switch t {
-	case "compute":
-		sc, err = rackspace.NewComputeV2(pc, gophercloud.EndpointOpts{
-			Region: region,
-		})
-		break
-	case "blockstorage":
-		sc, err = rackspace.NewBlockStorageV1(pc, gophercloud.EndpointOpts{
-			Region: region,
-		})
-		break
-	case "networking":
-		sc, err = rackspace.NewNetworkV2(pc, gophercloud.EndpointOpts{
-			Region: region,
-		})
-		break
-	}
-	if err != nil {
-		fmt.Printf("Error creating ServiceClient (%s): %s\n", err, t)
-		os.Exit(1)
+
+	// form the cache key
+	cacheKey := CacheKey(ao, region, serviceType)
+	// initialize cache
+	cache := &Cache{}
+	// get the value from the cache
+	creds, err := cache.Value(cacheKey)
+	// if there was an error accessing the cache or there was nothing in the cache,
+	// authenticate from scratch
+	if err == nil && creds != nil {
+		// we successfully retrieved a value from the cache
+		pc, err := rackspace.NewClient(ao.IdentityEndpoint)
+		if err == nil {
+			pc.TokenID = creds.TokenID
+			pc.ReauthFunc = reauthFunc(pc, ao)
+			pc.UserAgent.Prepend("rack/" + util.Version)
+			return &gophercloud.ServiceClient{
+				ProviderClient: pc,
+				Endpoint:       creds.ServiceEndpoint,
+			}, nil
```

Comment:
If we have to reauth, it seems like we should end up down in that else block, repopulating even the service endpoint (it could have changed after all).